File: gecko/bot/control_panel.py
"""Owner-only control panel: an embed + 8 buttons for everyday actions."""
from __future__ import annotations

import discord

from .auth import is_authorized_member
from .embeds import deny_embed
from .restock import do_check_tokens, do_cleanup_servers
from . import embeds as _embeds


def control_panel_embed() -> discord.Embed:
    return (
        discord.Embed(
            title="🛠️ Memberty Control Panel",
            description=(
                "Quick access to common owner actions. Buttons are owner-only "
                "and ephemeral — only the user who clicked sees the result."
            ),
            color=0x5865F2,
        )
        .set_footer(text="Memberty • Use the buttons below")
    )


def _btn(label: str, custom_id: str, *, style=discord.ButtonStyle.secondary, emoji=None):
    return discord.ui.Button(label=label, custom_id=custom_id, style=style, emoji=emoji)


def control_panel_view() -> discord.ui.View:
    """A persistent View — buttons keep working across restarts."""
    view = discord.ui.View(timeout=None)
    view.add_item(_btn("Stock", "cp:stock", emoji="📦"))
    view.add_item(_btn("Status", "cp:status", emoji="🟢"))
    view.add_item(_btn("Count", "cp:count", emoji="📊"))
    view.add_item(_btn("List Users", "cp:list_users", emoji="👥"))
    view.add_item(_btn("Check Tokens", "cp:check_tokens", emoji="🔍", style=discord.ButtonStyle.primary))
    view.add_item(_btn("Servers", "cp:servers", emoji="🌐"))
    view.add_item(_btn("Owners", "cp:owners", emoji="👑"))
    view.add_item(_btn("Cleanup Servers", "cp:cleanup", emoji="🧹", style=discord.ButtonStyle.danger))
    return view


async def handle_control_panel_button(
    interaction: discord.Interaction, server_join_times: dict
) -> None:
    """Dispatch cp:* button clicks. Always replies ephemerally."""
    cid = interaction.data.get("custom_id", "") if interaction.data else ""
    action = cid.split(":", 1)[1] if ":" in cid else ""
    guild = interaction.guild
    if guild is None:
        await interaction.response.send_message("❌ Guild only.", ephemeral=True)
        return
    guild_owner_id = str(guild.owner_id)
    user_id = str(interaction.user.id)
    if not is_authorized_member(guild_owner_id, str(guild.id), user_id, interaction.user):
        await interaction.response.send_message(embed=deny_embed(), ephemeral=True)
        return

    if action == "stock":
        await interaction.response.send_message(embed=_embeds.stock_embed(), ephemeral=True)
    elif action == "status":
        await interaction.response.send_message(
            embed=_embeds.status_embed(interaction.client, getattr(interaction.client, "_bot_start_time", None)),
            ephemeral=True,
        )
    elif action == "count":
        await interaction.response.send_message(embed=_embeds.count_embed(), ephemeral=True)
    elif action == "list_users":
        embed, _ = _embeds.list_users_embed()
        await interaction.response.send_message(embed=embed, ephemeral=True)
    elif action == "check_tokens":
        await interaction.response.defer(ephemeral=True)
        embed = await do_check_tokens()
        await interaction.followup.send(embed=embed, ephemeral=True)
    elif action == "servers":
        await interaction.response.send_message(
            embed=_embeds.servers_embed(interaction.client, server_join_times), ephemeral=True
        )
    elif action == "owners":
        await interaction.response.send_message(
            embed=_embeds.owners_embed(guild_owner_id, str(guild.id)), ephemeral=True
        )
    elif action == "cleanup":
        await interaction.response.defer(ephemeral=True)
        embed = await do_cleanup_servers(interaction.client, str(guild.id))
        await interaction.followup.send(embed=embed, ephemeral=True)
    else:
        await interaction.response.send_message("❌ Unknown action.", ephemeral=True)

File: src/bot/controlPanel.ts
import {
  ActionRowBuilder,
  ButtonBuilder,
  ButtonStyle,
  EmbedBuilder,
  type ButtonInteraction,
} from "discord.js";
import { COLOR } from "../config.js";
import { isAuthorizedMember } from "./permissions.js";
import { denyEmbed, listUsersEmbed, ownersEmbed, statusEmbed, stockEmbed, countEmbed, serversEmbed } from "./embeds.js";
import { doCheckTokens, doCleanupServers } from "./restock.js";
import type { BotState } from "./client.js";

export function controlPanelEmbed(): EmbedBuilder {
  return new EmbedBuilder()
    .setTitle("🛠️ Memberty Control Panel")
    .setDescription(
      "Quick access to common owner actions. Buttons are owner-only " +
        "and ephemeral — only the user who clicked sees the result.",
    )
    .setColor(COLOR.blurple)
    .setFooter({ text: "Memberty • Use the buttons below" });
}

function btn(
  label: string,
  customId: string,
  emoji: string,
  style: ButtonStyle = ButtonStyle.Secondary,
): ButtonBuilder {
  return new ButtonBuilder()
    .setLabel(label)
    .setCustomId(customId)
    .setEmoji(emoji)
    .setStyle(style);
}

export function controlPanelComponents(): ActionRowBuilder<ButtonBuilder>[] {
  const row1 = new ActionRowBuilder<ButtonBuilder>().addComponents(
    btn("Stock", "cp:stock", "📦"),
    btn("Status", "cp:status", "🟢"),
    btn("Count", "cp:count", "📊"),
    btn("List Users", "cp:list_users", "👥"),
  );
  const row2 = new ActionRowBuilder<ButtonBuilder>().addComponents(
    btn("Check Tokens", "cp:check_tokens", "🔍", ButtonStyle.Primary),
    btn("Servers", "cp:servers", "🌐"),
    btn("Owners", "cp:owners", "👑"),
    btn("Cleanup Servers", "cp:cleanup", "🧹", ButtonStyle.Danger),
  );
  return [row1, row2];
}

export async function handleControlPanelButton(
  interaction: ButtonInteraction,
  state: BotState,
): Promise<void> {
  const cid = interaction.customId;
  const action = cid.includes(":") ? cid.split(":", 2)[1] : "";
  const guild = interaction.guild;
  if (!guild) {
    await interaction.reply({ content: "❌ Guild only.", ephemeral: true });
    return;
  }
  const guildOwnerId = guild.ownerId;
  const userId = interaction.user.id;
  const member = interaction.member && "roles" in interaction.member
    ? (await guild.members.fetch(userId).catch(() => null))
    : null;
  if (!isAuthorizedMember(guildOwnerId, guild.id, userId, member)) {
    await interaction.reply({ embeds: [denyEmbed()], ephemeral: true });
    return;
  }

  switch (action) {
    case "stock":
      await interaction.reply({ embeds: [stockEmbed()], ephemeral: true });
      break;
    case "status":
      await interaction.reply({
        embeds: [statusEmbed(interaction.client, state.botStartTime)],
        ephemeral: true,
      });
      break;
    case "count":
      await interaction.reply({ embeds: [countEmbed()], ephemeral: true });
      break;
    case "list_users": {
      const { embed } = listUsersEmbed();
      await interaction.reply({ embeds: [embed], ephemeral: true });
      break;
    }
    case "check_tokens": {
      await interaction.deferReply({ ephemeral: true });
      const e = await doCheckTokens();
      await interaction.followUp({ embeds: [e], ephemeral: true });
      break;
    }
    case "servers":
      await interaction.reply({
        embeds: [serversEmbed(interaction.client, state.serverJoinTimes)],
        ephemeral: true,
      });
      break;
    case "owners":
      await interaction.reply({
        embeds: [ownersEmbed(guildOwnerId, guild.id)],
        ephemeral: true,
      });
      break;
    case "cleanup": {
      await interaction.deferReply({ ephemeral: true });
      const e = await doCleanupServers(interaction.client, guild.id);
      await interaction.followUp({ embeds: [e], ephemeral: true });
      break;
    }
    default:
      await interaction.reply({ content: "❌ Unknown action.", ephemeral: true });
  }
}
